text/0000-unsafe-fields.md

```diff
@@ -0,0 +1,1037 @@
+their additional complexity exceeds that of our proposed design. Our proposed rule that *a field
+marked `unsafe` is unsafe to use* is both pedagogically simple and fail safe; i.e., so long as a
+field is marked `unsafe`, it cannot be misused in such a way that its invariant is violated in safe
+code.
+
+### Mixing Syntactic Knobs with a Wrapper Type
+
+One alternative proposed in this RFC's discussion recommends a combination of syntactic knobs and a
+wrapper type. In brief, a simple [`Unsafe` wrapper type](#unsafe-wrapper-type) would be provided,
+along with two field safety modifiers:
+
+- `unsafe`
+  All uses except reading are `unsafe`.
+- `unsafe(mut)`
+  All uses except reading and `&`-referencing are `unsafe`.
+
+Under this proposal, a programmer would use some combination of `unsafe`, `unsafe(mut)` and `Unsafe`
+to precisely tune Rust's safety tooling protections, depending on the hazards of their invariant.
+
+The primary advantage of this approach is that it results in comparatively fewer instances in which
+[the programmer must write a 'trivial' safety proof](#trivial-safety-proofs). However, it achieves
+this by front-loading the requirement that the programmer imagine all possible safety hazards of
+their field. A mistake, here, may lead to a false sense of security if Rust fails to require
+`unsafe` for uses that are, in fact, dangerous. By contrast, this RFC requires that programmers
+resolve these questions only on an as-needed basis; e.g., until you need to `&`-reference a field,
+you do not need to confront whether doing so is *always* a safe operation.
+
+This alternative also inherits some of the disadvantages of [`Unsafe` wrapper
+types](#unsafe-wrapper-type); namely that the safety proofs needed to operate on an `Unsafe` wrapper
+value depend on the dataflow of the program; the wrapper value must be traced to its originating
+field so that field's safety documentation may be examined.
+
+Comparatively, we believe that this RFC's proposal is both pedagogically simpler and less prone to
+misuse, and that these benefits outweigh its [drawbacks](#drawbacks).
+
+# Drawbacks
+
+## Trivial Safety Proofs
+
+The primary drawback of this proposal is that it — in some scenarios — necessitates writing
+'trivial' safety proofs. For example, merely reading `Vec`'s `len` field obviously cannot invalidate
+its invariant; nonetheless, this field, if marked `unsafe`, would be `unsafe` to read. An `unsafe`
+block and attendant `SAFETY` comment is required. In most cases, this is a one-time chore: the
+maintainer can define a *safe* accessor (i.e.,
+[`Vec::len`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.len)) that encapsulates this
+proof. However, in cases where multiple, partial field borrows are required, such an accessor cannot
+be invoked. [Future language extensions that permit partial borrows may resolve this
+drawback.](#partial-borrows).
```

Comment:
Even with partial borrows, it may not be possible to write a safe helper method for *moving out* of a field (while leaving the rest of the struct intact).